overlord/snapstate/sequence/sequence.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// sequence contains types describing a sequence of snap revisions
+// (with components) that describe current and past states of the snap
+// in the system.
+package sequence
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+// RevisionSideState contains the side information for a snap and related components
+// installed in the system.
+type RevisionSideState struct {
+	Snap       *snap.SideInfo
+	Components []*snap.ComponentSideInfo
+}
+
+// revisionSideInfoMarshal is an ancillary structure used exclusively to
+// help marshaling of RevisionSideInfo.
+type revisionSideInfoMarshal struct {
+	// SideInfo is included for compatibility with older snapd state files
+	*snap.SideInfo
+	CompSideInfo []*snap.ComponentSideInfo `yaml:"components,omitempty" json:"components,omitempty"`
+}
+
+// MarshalJSON implements the json.Marshaler interface
+func (bsi RevisionSideState) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&revisionSideInfoMarshal{bsi.Snap, bsi.Components})
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+func (bsi *RevisionSideState) UnmarshalJSON(in []byte) error {
```

Comment:
this still seems untested here

Reply:
Right, fixed now, not sure how I overlooked this